hack/verify-release.sh

```diff
@@ -0,0 +1,587 @@
+        if ! git ls-remote --exit-code --tags "${REMOTE}" "refs/tags/v${VERSION}" &>/dev/null; then
+            echo "ERROR: tag ${tag} is not found in remote ${REMOTE}"
+        fi
+    done
+
+    echo -e "Done\n"
+}
+
+verify_git_tag_types()
+{
+    # check tags are annotated or lightweight as expected
+    # and also that no extra tags are pushed by accident
+    echo "Verifying Git tag types ..."
+
+    for annotated_tag in "${git_annotated_tags[@]}"; do
+        if [[ "$(git cat-file -t "${annotated_tag}" 2>/dev/null)" != "tag" ]]; then
+            echo "ERROR: ${annotated_tag} is not an annotated tag, or is missing"
+        fi
+    done
+
+    for lightweight_tag in "${git_lightweight_tags[@]}"; do
+        if [[ "$(git cat-file -t "${lightweight_tag}" 2>/dev/null)" != "commit" ]]; then
+            echo "WARNING: ${lightweight_tag} is not a lightweight tag, or is missing"
+        fi
+    done
+
+    for nonexist_tag in "${git_nonexisting_tags[@]}"; do
+        if git cat-file -t "${nonexist_tag}" &>/dev/null; then
+            echo "ERROR: ${nonexist_tag} is exists, while it should not"
+        fi
+    done
+
+    echo -e "Done\n"
+}
+
+verify_release_notes()
+{
+    # check release note content
+    echo "Verifying release notes ..."
+
+    # check body if certain strings
+    for string in "${release_note_tag_strings[@]}"; do
+        # shellcheck disable=SC2076
+        if ! [[ "$(jq .body "${RELEASE_JSON}")" =~ "${string}" ]]; then
+            echo "ERROR: '${string}' not found in release note text, is tag correct?"
+        fi
+    done
+
+    # check body for tagged images
+    for string in "${release_note_strings[@]}"; do
+        # shellcheck disable=SC2076
+        if ! [[ "$(jq .body "${RELEASE_JSON}")" =~ "${string}" ]]; then
+            echo "WARNING: '${string}' not found in release note text, recheck content"
+        fi
+    done
+
+    echo -e "Done\n"
+}
+
+verify_release_artefacts()
+{
+    # check that the release json lists all artefacts as present
+    echo "Verifying release artefacts ..."
+
+    for artefact in "${release_artefacts[@]}"; do
+        # shellcheck disable=SC2076
+        if ! [[ "$(jq .assets[].name "${RELEASE_JSON}")" =~ "\"${artefact}\"" ]]; then
+            echo "ERROR: release artefact '${artefact}' not found in release"
+        fi
+    done
+
+    echo -e "Done\n"
+}
+
+verify_container_images()
+{
+    # check quay as built images successfully, and hence tag is present
+    # if tag doesn't appear, the build trigger might've been disabled
+    local image tag
+
+    echo "Verifying container images ..."
+
+    for image_and_tag in "${container_images[@]}"; do
+        image="${image_and_tag/:*}"
+        tag="${image_and_tag/*:}"
+
+        # quay paginates 50 items at a time, so it is simpler to use gcrane
+        # to list all the tags, than DIY parse the pagination logic
+        if ! "${CONTAINER_RUNTIME}" run --rm gcr.io/go-containerregistry/gcrane \
```

Comment:
Crane and osv-scanner are now executed via Docker, maybe I could just expect user to have them installed via "go install" instead. Docker is OK for Linux, but Mac users might have an issue with it.

/hold

Reply:
OK. I changed it so that if `CONTAINER_RUNTIME` is defined, it will run gcrane and osv-scanner as images, otherwise it will check for pre-installed binaries. This caters for most users.

/unhold